finagle-base-http/src/main/scala/com/twitter/finagle/http/headers/JTreeMapBackedHeaderMap.scala

```diff
@@ -0,0 +1,139 @@
+package com.twitter.finagle.http.headers
+
+import com.twitter.finagle.http.HeaderMap
+import scala.annotation.tailrec
+import java.util.function.BiConsumer
+import scala.collection.AbstractIterator
+
+/**
+ * Mutable, thread-safe [[HeaderMap]] implementation, backed by
+ * a mutable [[Map[String, Header]]], where the map key
+ * is forced to lower case
+ */
+final private class JTreeMapBackedHeaderMap extends HeaderMap {
+  import HeaderMap._
+
+  // In general, Map's that are not thread safe are not
+  // durable to concurrent modification and can result in infinite loops
+  // and exceptions.
+  // As such, we synchronize on the underlying collection when performing
+  // accesses to avoid this. In the common case of no concurrent access,
+  // this should be cheap.
+  val underlying: java.util.TreeMap[String, Header.Root] = 
+  new java.util.TreeMap[String, Header.Root](JTreeMapBackedHeaderMap.SharedComparitor)
+
+  private def foreachConsumer[U](f: ((String, String)) => U):
+    BiConsumer[String, Header.Root] = new BiConsumer[String, Header.Root](){
+      def accept(key: String, header: Header.Root): Unit = header.iterator.foreach(
+        nv => f(nv.name, nv.value)
+      )
+    }
+
+  override def foreach[U](f: ((String, String)) => U): Unit = 
```

Comment:
Now I wish I never removed the abstraction to swap out the underlying Map implementation. 

But let's take a step back. How often do people really concurrently iterate over headers and also mutate the header map and don't synchronize that access because they don't care whether or not the modified data should or shouldn't be part of the iteration? Is that a use-case you want to facilitate in the first place?

Thinking about those questions, my conclusion is that throwing a ConcurrentModificationException that makes the user aware of the situation/race condition/potential bug is better than synchronizing and not being deterministic in what happens first. If the user *really* wants that for some unforeseen, unspoken and unspeakable reason, they can always still externally synchronize (maybe with a ReadWriteLock or something).

I think in terms of common usage patterns, that's the fastest and most reasonable thing to do too. Even if a ConcurrentSkipListMap would be pretty cool.

If we do want to get overboard and fancy, it should be validated with benchmarks that reflect real-world usage patterns, including the real-world concurrent patterns, because otherwise we're not really measuring anything close to reality.

Reply:
I think the *good* outcome is that they get a ConcurrentModificationException–a bad outcome might be an infinite loop or worse.  See https://ivoanjo.me/blog/2018/07/21/writing-to-a-java-treemap-concurrently-can-lead-to-an-infinite-loop-during-reads/

I think correctness needs to be our #1 concern.  I agree with you that it's extremely unlikely–that's one of the reasons I'm willing to just wrap it in a synchronized block and call it a day.

I'm not sure I understand your concern about not being deterministic about what happens first–that's always going to be the case with concurrent use of a data structure.  My concern is that HeaderMap used to be threadsafe, so people may have been already using it assuming it was threadsafe, and then one day we merge this in and their application breaks in an extremely subtle way.  If this was a new abstraction, I would be open to considering, "HeaderMap is not threadsafe, all concurrent access must be synchronized by the user" but I don't think that's the right way forward given the circumstances.

My bet is the same as yours, that it's extremely unusual to access it concurrently, so I don't think we need to benchmark it.  But I do think that we should protect against it.  Locking is pretty cheap when there's no concurrent access, so I don't think it should be that painful to just lock it.

Reply:
Jikes! I didn't realize the concurrent case could lead to non-termination. That makes not synchronizing a non-starter, and the below just an aside for academic curiosity. 

The concern about concurrent iteration and modification (specifically those two) is that the concept of iteration of all headers and concurrently modifying those headers makes no sense -- either you want those in or out, but you always care, or you wouldn't iteratate them. You must synchronise externally for the operation to be meaningful and correct.

That's not always the case for two concurrent modifications that will never modify the same key, or concurrently getting a key, and modifying some keys that you know to be different to the keys you're concurrently getting.

Reply:
I've run into this non-termination before... It was a nightmare, especially considering it was happening on a Netty thread.

Normally I'd agree that we'd prefer a `ConcurrentModificationException` and we've tried that in the past but many users are doing the wrong thing in benign places (think reusing a request to do warmup requests or something) and it just wasn't worth the hassle to introduce the behavior change.

I think the solution you've already committed of simply synchronizing the `foreach` call is fine since you're right: this shouldn't be happening very often, if at all. 🤞